lib/ide/json-rpc/src/handler.rs

```diff
@@ -150,55 +134,141 @@ pub type OngoingCalls = HashMap<Id,oneshot::Sender<ReplyMessage>>;
 /// `Notification` is a type for notifications. It should implement
 /// `DeserializeOwned` and deserialize from JSON maps with `method` and `params`
 /// fields.
+///
 #[derive(Debug)]
 pub struct Handler<Notification> {
-    /// Contains handles to calls that were made but no response has came.
-    pub ongoing_calls   : OngoingCalls,
-    /// Provides identifiers for requests.
-    pub id_generator    : IdGenerator,
-    /// Transports text messages between this handler and the peer.
-    pub transport       : Box<dyn Transport>,
-    /// Allows receiving events from the `Transport`.
-    pub incoming_events : std::sync::mpsc::Receiver<TransportEvent>,
-    /// Handle to send outgoing events.
-    pub outgoing_events : Option<UnboundedSender<Event<Notification>>>,
+    rc: Rc<RefCell<HandlerData<Notification>>>,
 }
 
+
+// === HandlerData accessors ===
+
+impl<Notification> Handler<Notification> {
+    /// Inserts a new entry for an ongoing request awaiting reply.
+    pub fn insert_ongoing_request(&self, id:Id, sender:oneshot::Sender<ReplyMessage>) {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.insert(id,sender);
+        })
+    }
+
+    /// Removes the request from the map of ongoing requests, e.g. because it
+    /// has been successfully completed or we know that we will not receive an
+    /// answer anymore.
+    ///
+    /// Returns the channel handle for the request, it should be immediately
+    /// after notified or dropped.
+    pub fn remove_ongoing_request(&self, id:Id) -> Option<oneshot::Sender<ReplyMessage>> {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.remove(&id)
+        })
+    }
+
+    /// Removes all the ongoing requests. This will be recognized by the `Future`s
+    /// as losing connection error.
+    pub fn clear_ongoing_requests(&self) {
+        with(self.rc.borrow_mut(), |mut data| {
+            data.ongoing_calls.clear()
+        })
+    }
+
```

Comment:
Instead of doing all of these things by hand,  could you pelase use the `shared!` macro?